src/main/java/com/powsybl/openloadflow/ac/VoltageMagnitudeInitializer.java

```diff
@@ -0,0 +1,339 @@
+                    if (piModel.getX() == 0) { // skip zero impedance branches
+                        continue;
+                    }
+                    b += Math.abs(1 / piModel.getX()); // to void issue with negative reactances
+                    r += neighborBranch.getBus1() == bus ? 1 / piModel.getR1() : piModel.getR1();
+                    neighborCount++;
+                }
+                if (neighborCount > 0) {
+                    r /= neighborCount;
+
+                    bs += b;
+                    der.add(b * r);
+
+                    // add variable
+                    variables.add(variableSet.getVariable(neighborBus.getNum(), InitVmVariableType.BUS_V));
+                }
+            }
+            if (bs == 0) { // should never happen
+                throw new PowsyblException("Susceptance sum is zero");
+            }
+            for (int i = 0; i < der.size(); i++) {
+                der.setQuick(i, der.getQuick(i) / bs);
+            }
+        }
+
+        @Override
+        public List<Variable<InitVmVariableType>> getVariables() {
+            return variables;
+        }
+
+        @Override
+        public void update(double[] x) {
+            // nothing that depends on state
+        }
+
+        @Override
+        public double eval() {
+            throw new IllegalStateException("Useless");
+        }
+
+        @Override
+        public double der(Variable<InitVmVariableType> variable) {
+            int i = variables.indexOf(variable);
+            if (i == -1) {
+                throw new IllegalStateException("Unknown variable: " + variable);
+            }
+            return der.getQuick(i);
+        }
+
+        @Override
+        protected String getName() {
+            return "v_distr";
+        }
+    }
+
+    private final MatrixFactory matrixFactory;
+
+    public VoltageMagnitudeInitializer(MatrixFactory matrixFactory) {
+        this.matrixFactory = Objects.requireNonNull(matrixFactory);
+    }
+
+    private static void initTarget(Equation<InitVmVariableType, InitVmEquationType> equation, LfNetwork network, double[] targets) {
+        switch (equation.getType()) {
+            case BUS_TARGET_V:
+                int busNum = equation.getData() == null ? equation.getNum() : (Integer) equation.getData();
+                LfBus bus = network.getBus(busNum);
+                targets[equation.getColumn()] = bus.getVoltageControl().orElseThrow().getTargetValue();
+                break;
+            case BUS_ZERO:
+            case BRANCH_ZERO:
+                targets[equation.getColumn()] = 0;
+                break;
+            default:
+                throw new IllegalStateException("Unknown equation type: " + equation.getType());
+        }
+    }
+
+    private static boolean isZeroImpedanceBranch(LfBranch branch) {
+        return branch.getBus1() != null
+                && branch.getBus2() != null
+                && branch.getPiModel().getX() == 0;
```

Comment:
My first think here is to use better `branch.isSpanningTreeEdge()` to avoid cycle. It is what we have done for the AC equation system.